File: app/src/main/java/com/raival/fileexplorer/tab/file/dialog/FileInfoDialog.java
package com.raival.fileexplorer.tab.file.dialog;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.raival.fileexplorer.App;
import com.raival.fileexplorer.R;
import com.raival.fileexplorer.util.FileUtils;
import com.raival.fileexplorer.util.Utils;

import java.io.File;
import java.util.ArrayList;

public class FileInfoDialog extends BottomSheetDialogFragment {
    private final ArrayList<InfoHolder> infoList = new ArrayList<>();
    private final File file;
    private boolean useDefaultFileInfo;
    private ViewGroup container;

    public FileInfoDialog(@NonNull File file) {
        this.file = file;
    }

    public FileInfoDialog setUseDefaultFileInfo(boolean is) {
        useDefaultFileInfo = is;
        return this;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.file_explorer_tab_info_dialog, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ((TextView) view.findViewById(R.id.file_name)).setText(file.getName());
        FileUtils.setFileIcon(view.findViewById(R.id.file_icon), file);

        container = view.findViewById(R.id.container);

        if (!useDefaultFileInfo) {
            for (InfoHolder holder : infoList) {
                addItemView(holder, container);
            }
        } else {
            if (file.isFile()) {
                addDefaultFileInfo();
            } else {
                addDefaultFolderInfo();
            }
        }

    }

    private void addDefaultFolderInfo() {
        addItemView(new InfoHolder("Path:", file.getAbsolutePath(), true), container);
        addItemView(new InfoHolder("Modified:", Utils.getLastModifiedDate(file), true), container);
        addItemView(new InfoHolder("Content:", FileUtils.getFormattedFileCount(file), true), container);
        addItemView(new InfoHolder("Type:", file.isFile() ? "File" : "Folder", true), container);
        addItemView(new InfoHolder("Read:", file.canRead() ? "Yes" : "No", true), container);
        addItemView(new InfoHolder("Write:", file.canWrite() ? "Yes" : "No", true), container);

        TextView size = addItemView(new InfoHolder("Size:", "Counting...", true), container);
        new Thread(() -> {
            String s = FileUtils.getFormattedFileSize(file);
            size.post(() -> size.setText(s));
        }).start();
    }

    private void addDefaultFileInfo() {
        addItemView(new InfoHolder("Path:", file.getAbsolutePath(), true), container);
        addItemView(new InfoHolder("Extension:", FileUtils.getFileExtension(file), true), container);
        addItemView(new InfoHolder("Modified:", Utils.getLastModifiedDate(file), true), container);
        addItemView(new InfoHolder("Type:", file.isFile() ? "File" : "Folder", true), container);
        addItemView(new InfoHolder("Read:", file.canRead() ? "Yes" : "No", true), container);
        addItemView(new InfoHolder("Write:", file.canWrite() ? "Yes" : "No", true), container);
        addItemView(new InfoHolder("Size:", FileUtils.getFormattedFileSize(file), true), container);
    }

    private TextView addItemView(InfoHolder holder, ViewGroup container) {
        @SuppressLint("InflateParams") View view = getLayoutInflater().inflate(R.layout.file_explorer_tab_info_dialog_item, null, false);

        ((TextView) view.findViewById(R.id.name)).setText(holder.name);
        TextView details = view.findViewById(R.id.details);
        details.setText(holder.info);
        if (holder.clickable) {
            details.setClickable(true);
            details.setOnClickListener(view1 -> {
                App.copyString(holder.info);
                App.showMsg(holder.name + " has been copied");
            });
        }
        container.addView(view);
        return details;
    }

    public FileInfoDialog addItem(String name, String info, boolean clickable) {
        infoList.add(new InfoHolder(name, info, clickable));
        return this;
    }

    @Override
    public int getTheme() {
        return R.style.ThemeOverlay_Material3_BottomSheetDialog;
    }

    public static class InfoHolder {
        public String name;
        public String info;
        public boolean clickable;

        public InfoHolder(String name, String info, boolean clickable) {
            this.name = name;
            this.info = info;
            this.clickable = clickable;
        }
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/dialog/FileInfoDialog.kt
package com.raival.fileexplorer.tab.file.dialog

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.textfield.TextInputLayout
import com.raival.fileexplorer.R
import com.raival.fileexplorer.extension.getFolderSize
import com.raival.fileexplorer.extension.getFormattedFileCount
import com.raival.fileexplorer.extension.getLastModifiedDate
import com.raival.fileexplorer.extension.toFormattedSize
import com.raival.fileexplorer.tab.file.misc.IconHelper
import com.raival.fileexplorer.tab.file.misc.md5.HashUtils
import com.raival.fileexplorer.tab.file.misc.md5.MessageDigestAlgorithm
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.security.MessageDigest

class FileInfoDialog(private val file: File) : BottomSheetDialogFragment() {
    private val infoList = ArrayList<InfoHolder>()
    private var useDefaultFileInfo = false
    private lateinit var container: ViewGroup

    fun setUseDefaultFileInfo(`is`: Boolean): FileInfoDialog {
        useDefaultFileInfo = `is`
        return this
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.file_explorer_tab_info_dialog, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (view.findViewById<View>(R.id.file_name) as TextView).text = file.name
        IconHelper.setFileIcon(view.findViewById(R.id.file_icon), file)
        container = view.findViewById(R.id.container)
        if (!useDefaultFileInfo) {
            for (holder in infoList) {
                addItemView(holder, container)
            }
        } else {
            if (file.isFile) {
                addDefaultFileInfo()
            } else {
                addDefaultFolderInfo()
            }
        }
    }

    private fun addDefaultFolderInfo() {
        addItemView(InfoHolder("Path:", file.absolutePath, true), container)
        addItemView(
            InfoHolder(
                "Modified:", file.getLastModifiedDate(), true
            ), container
        )
        val count = addItemView(InfoHolder("Content:", "Counting...", true), container)
        CoroutineScope(Dispatchers.IO).launch {
            val s = file.getFormattedFileCount()
            withContext(Dispatchers.Main) { count.text = s }
        }.start()
        val size = addItemView(InfoHolder("Size:", "Counting...", true), container)
        CoroutineScope(Dispatchers.IO).launch {
            val s = file.getFolderSize().toFormattedSize()
            withContext(Dispatchers.Main) { size.text = s }
        }.start()
    }

    private fun addDefaultFileInfo() {
        addItemView(InfoHolder("Path:", file.absolutePath, true), container)
        addItemView(
            InfoHolder(
                "Extension:", file.extension, true
            ), container
        )
        addItemView(
            InfoHolder(
                "Modified:", file.getLastModifiedDate(), true
            ), container
        )
        addItemView(
            InfoHolder(
                "Size:", file.length().toFormattedSize(), true
            ), container
        )

        val md5 = addItemView(InfoHolder("MD5:", "calculating...", true), container)
        CoroutineScope(Dispatchers.IO).launch {
            val s = HashUtils.getCheckSumFromFile(
                MessageDigest.getInstance(MessageDigestAlgorithm.MD5),
                file
            )
            withContext(Dispatchers.Main) { md5.text = s }
        }.start()

        val sha1 = addItemView(InfoHolder("SHA1:", "calculating...", true), container)
        CoroutineScope(Dispatchers.IO).launch {
            val s = HashUtils.getCheckSumFromFile(
                MessageDigest.getInstance(MessageDigestAlgorithm.SHA_1),
                file
            )
            withContext(Dispatchers.Main) { sha1.text = s }
        }.start()

        val sha256 = addItemView(InfoHolder("SHA256:", "calculating...", true), container)
        CoroutineScope(Dispatchers.IO).launch {
            val s = HashUtils.getCheckSumFromFile(
                MessageDigest.getInstance(MessageDigestAlgorithm.SHA_256),
                file
            )
            withContext(Dispatchers.Main) { sha256.text = s }
        }.start()

        val sha512 = addItemView(InfoHolder("SHA_512:", "calculating...", true), container)
        CoroutineScope(Dispatchers.IO).launch {
            val s = HashUtils.getCheckSumFromFile(
                MessageDigest.getInstance(MessageDigestAlgorithm.SHA_512),
                file
            )
            withContext(Dispatchers.Main) { sha512.text = s }
        }.start()
    }

    private fun addItemView(holder: InfoHolder, container: ViewGroup?): TextView {
        @SuppressLint("InflateParams") val view =
            layoutInflater.inflate(R.layout.file_explorer_tab_info_dialog_item, null, false)
        val v = (view.findViewById<View>(R.id.item) as TextInputLayout).apply {
            editText!!.apply {
                keyListener = null
                setTextIsSelectable(true)
                setText(holder.info)
            }
            hint = holder.name
        }
        container!!.addView(view)
        return v.editText!!
    }

    fun addItem(name: String, info: String?, clickable: Boolean): FileInfoDialog {
        infoList.add(InfoHolder(name, info, clickable))
        return this
    }

    override fun getTheme(): Int {
        return R.style.ThemeOverlay_Material3_BottomSheetDialog
    }

    class InfoHolder(var name: String, var info: String?, var clickable: Boolean)
}